WeatherApp/WeatherApp/CityListScreen/CityListViewModel.swift

```diff
@@ -2,45 +2,58 @@ import SwiftUI
 
 class CityListViewModel: ObservableObject {
 
-    @Published var cities: [City] = [
+    @Published private(set) var cities: [City] = [
         City(name: "Zagreb"),
         City(name: "Paris"),
         City(name: "New York"),
         City(name: "Tokyo"),
         City(name: "London"),
         City(name: "Los Angeles"),
         City(name: "Toronto"),
-        City(name: "Split")
-    ]
+        City(name: "Split")]
 
     private let router: RouterProtocol
     private let weatherService: WeatherServiceProtocol
 
     init(router: RouterProtocol, service: WeatherServiceProtocol) {
         self.router = router
         self.weatherService = service
+
+        fetchWeatherForAllCities()
     }
 
-    func fetchTemperature(for city: City) async {
-            do {
-                let weather = try await weatherService.fetchWeather(for: city.name)
-                if let index = cities.firstIndex(where: { $0.id == city.id }) {
-                    DispatchQueue.main.async {
-                        self.cities[index].temperature = weather.main.temp
+    func fetchTemperature(for city: City) {
+        weatherService.fetchWeather(for: city.name) { [weak self] result in
+            guard let self else { return }
+
+            switch result {
+            case .success(let weatherResponse):
+                if let index = self.cities.firstIndex(where: { $0.id == city.id }),
+                   self.cities.at(index) != nil {
```

Comment:
also here you are only checking if its not nil and in the DispatchQueue you are using subscript to extract the city for index.

Proper approach would be to `let` this value and use it moving forward